packages/shared/ReactMutableSource.js

```diff
@@ -0,0 +1,134 @@
+export type Version = $NonMaybeType<mixed>;
+
+// Tracks expiration time for all mutable sources with pending updates.
+// Used to determine if a source is safe to read during updates.
+// If there are no entries in this map for a given source,
+// or if the current render’s expiration time is ≤ this value,
+// it is safe to read from the source without tearing.
+export type MutableSourcePendingUpdateMap = Map<
+  MutableSource<any>,
+  ExpirationTime,
+>;
+
+type MutableSourceConfig = {|
+  getVersion: () => Version,
+|};
+
+export type MutableSource<Source: $NonMaybeType<mixed>> = {|
+  _source: Source,
+
+  _getVersion: () => Version,
+
+  // Tracks the version of this source at the time it was most recently read.
+  // Used to determine if a source is safe to read from before it has been subscribed to.
+  // Version number is only used during mount,
+  // since the mechanism for determining safety after subscription is expiration time.
+  //
+  // As a workaround to support multiple concurrent renderers,
+  // we categorize some renderers as primary and others as secondary.
+  // We only expect there to be two concurrent renderers at most:
+  // React Native (primary) and Fabric (secondary);
+  // React DOM (primary) and React ART (secondary).
+  // Secondary renderers store their context values on separate fields.
+  // We use the same approach for Context.
+  _workInProgressVersionPrimary: null | Version,
+  _workInProgressVersionSecondary: null | Version,
+|};
+
+export type MutableSourceHookConfig<Source: $NonMaybeType<mixed>, Snapshot> = {|
+  getSnapshot: (source: Source) => Snapshot,
+  subscribe: (source: Source, callback: Function) => () => void,
+|};
+
+// Work in progress version numbers only apply to a single render,
+// and should be reset before starting a new render.
+// This tracks which mutable sources need to be reset after a render.
+let workInProgressPrimarySources: Array<MutableSource<any>> = [];
+let workInProgressSecondarySources: Array<MutableSource<any>> = [];
+
+export function createMutableSource<Source: $NonMaybeType<mixed>>(
+  source: Source,
+  config: MutableSourceConfig,
+): MutableSource<Source> {
+  return {
+    _getVersion: config.getVersion,
+    _source: source,
+    _workInProgressVersionPrimary: null,
+    _workInProgressVersionSecondary: null,
+  };
+}
+
+export function clearPendingUpdates(
+  root: FiberRoot,
+  expirationTime: ExpirationTime,
+): void {
+  // Remove pending mutable source entries that we've completed processing.
+  root.mutableSourcePendingUpdateMap.forEach(
+    (pendingExpirationTime, mutableSource) => {
+      if (pendingExpirationTime <= expirationTime) {
+        root.mutableSourcePendingUpdateMap.delete(mutableSource);
+      }
+    },
+  );
+}
+
+export function resetWorkInProgressVersions(isPrimaryRenderer: boolean): void {
+  if (isPrimaryRenderer) {
+    workInProgressPrimarySources.forEach(mutableSource => {
```

Comment:
Nit: This is an array, not a Map or a Set, so let's use a for loop like we do elsewhere.